### PR TITLE
APEXCORE-421 #comment #resolve Fixing double checkpoint bug on InputN…

### PR DESCRIPTION
**Pull request by ilooner**
…ode shutdown

### CODE REVIEW

**Comment by ilooner** [comment]
@davidyan74 please review. There was another double checkpointing bug in InputNode which was causing my unit test to fail on travis.

**Comment by sandeshh** [comment]
If the checkpoint function is not properly used in 3 places, it is time to fix that as well.

**Comment by ilooner** [comment]
@sandeshh I agree, the problem is that this code is quite complex and there is almost no unit test coverage. I am afraid to do refactoring on code I don't completely understand, which is not sufficiently tested.

**Comment by ilooner** [comment]
@davidyan74 please review and merge

**Comment by ilooner** [comment]
@chandnisingh please review and merge

File: engine/src/main/java/com/datatorrent/stram/engine/InputNode.java [hunk 1]
```diff
@@ -241,11 +241,14 @@ else if (PROCESSING_MODE == ProcessingMode.EXACTLY_ONCE) {
       if (++applicationWindowCount == APPLICATION_WINDOW_COUNT) {
         applicationWindowCount = 0;
       }
-      if (++checkpointWindowCount == CHECKPOINT_WINDOW_COUNT) {
-        checkpointWindowCount = 0;
-        if (doCheckpoint || PROCESSING_MODE == ProcessingMode.EXACTLY_ONCE) {
-          checkpoint(currentWindowId);
-          lastCheckpointWindowId = currentWindowId;
+
+      if (lastCheckpointWindowId < currentWindowId) {
```
Comment (otterc) [inline engine/src/main/java/com/datatorrent/stram/engine/InputNode.java:245]: can we add a small comment here that why this check is needed?
It helps others to understand

**Comment by ilooner** [comment]
@chandnisingh added a comment about why the check is there.

**Pull request merged**